cclunit-framework-source/src/test/ccl/ut_cclut_mock.inc

```diff
@@ -0,0 +1,409 @@
+*  Scenario: Throws an error when an empty programName is supplied to the function (after trimming)         *
+************************************************************************************************************/
+subroutine test_cclutExecuteProgramWithMocks_missing_programName(null)
+	call echo("~~~***~~~***test_cclutExecuteProgramWithMocks_missing_programName***~~~***~~~")
+
+	declare errorMessage = vc with protect, noconstant("")
+    declare errorCode = i4 with protect, noconstant(0)
+	call cclutExecuteProgramWithMocks("     ", "")
+
+	set errorCode = error(errorMessage, 0)
+
+    call cclutAssertVcOperator(CURREF, "test_cclutExecuteProgramWithMocks_missing_programName", trim(errorMessage, 3),
+        "regexplike", concat("%CCL-E-392-PRG_[0-9]+_[0-9]+\([^)]+\)[0-9]+:[0-9]+\{CCLEXCEPTION\(\)\}Exception\(100\):",
+            " cclutExecuteProgramWithMocks\(\) - programName is a required field."))
+end ;test_cclutExecuteProgramWithMocks_missing_programName
+
+;**********************************************************************************************************************************
+;** cclutRemoveAllMocks
+;**********************************************************************************************************************************
+/* test_cclutRemoveAllMocks_happy ***************************************************************************
+*  Scenario: Removes all mock tables and implementations so that the originals can be used                  *
+************************************************************************************************************/
+subroutine test_cclutRemoveAllMocks_happy(null)
+	call echo("~~~***~~~***test_cclutRemoveAllMocks_happy***~~~***~~~")
+
+	call cclutDefineMockTable("sample_table", "sample_table_id|sample_table_text|sample_table_date", "f8|vc|dq8")
+    call cclutDefineMockTable("sample_table_2", "sample_table_2_id|sample_table_2_text|sample_table_2_date",
+        "f8|vc|dq8")
+    call cclutDefineMockTable("sample_table_3", "sample_table_3_id|sample_table_3_text|sample_table_3_date",
+        "f8|vc|dq8")
+
+    call cclutCreateMockTable("sample_table")
+
+	call cclutAddMockImplementation("sample_subroutine", "mock_subroutine")
+	call cclutAddMockImplementation("sample_subroutine_2", "mock_subroutine_2")
+	call cclutAddMockImplementation("sample_subroutine_3", "mock_subroutine_3")
+
+    call cclutAsserti4Equal(CURREF, "test_cclutRemoveAllMocks_happy",
+        size(cclut_mockTables->tables, 5), 3)
+    call cclutAsserti4Equal(CURREF, "test_cclutRemoveAllMocks_happy",
+        size(cclut_mockImplementations->implementations, 5), 3)
+
+    call cclutRemoveAllMocks(null)
+
+    call cclutAsserti4Equal(CURREF, "test_cclutRemoveAllMocks_happy", size(cclut_mockTables->tables, 5), 0)
+    call cclutAsserti4Equal(CURREF, "test_cclutRemoveAllMocks_happy",
+        size(cclut_mockImplementations->implementations, 5), 0)
+end ;test_cclutRemoveAllMocks_happy
```

Comment:
I think this is a great idea.  This also made me realize that I'm missing a use-case for the fail-fast scenario.

So, one, I need to add the removal there.  Then there should be two corresponding tests; one for the fail-fast and one for the regular.  I'm thinking for the fail-fast, since there's already a ut_cclut_ff.prg, that a misbehaving test case is created for that in src/test/resources, and then a new test is added to ut_cclut_ff that calls that and does the other things you mention (misbehaving checks it's there in teardownOnce and ut_cclut_ff checks that it's gone).  Since it's just calling down, the misbehaving script could set the name of the mock table that it gets back on to a variable defined by the test script.

For the regular one, I'm thinking it would do a similar thing (using the same misbehaving .inc), but from ut_cclut_execute_test_case(and file) instead.  However, in both these scenarios, the misbehaving test case _is_ in src/test/resources (whereas you made it sound like you wanted it in src/test/ccl), so I just want to make sure that this approach would still be in line with your expectations.  If not, can you provide some more insight into your thoughts?